### PR TITLE
Remove support for OTEL spanmetrics processor

### PR DESCRIPTION
**Pull request by varshith257**
## Which problem is this PR solving?
- Part of #4736 

## Description of the changes
-   removed support for the spanmetrics processor flag.

## How was this change tested?
-  CI

## Checklist
- [x] I have read https://github.com/jaegertracing/jaeger/blob/master/CONTRIBUTING_GUIDELINES.md
- [x] I have signed all commits
- [x] I have added unit tests for the new functionality
- [x] I have run lint and test steps successfully
  - for `jaeger`: `make lint test`
  - for `jaeger-ui`: `yarn lint` and `yarn test`

### CODE REVIEW

**Comment by varshith257** [comment]
@yurishkuro Still there any to remove? These are files mainly updated in the PR linked in the ticket for dual support of the processor and connector.

And for README, can we keep the migration guide or remove it?

File: plugin/metrics/prometheus/options.go [hunk 1]
```diff
@@ -97,7 +94,7 @@ func (opt *Options) AddFlags(flagSet *flag.FlagSet) {
 	flagSet.Bool(
 		nsConfig.namespace+suffixSupportSpanmetricsConnector,
```
Comment (albertteoh) [inline plugin/metrics/prometheus/options.go:95]: We should be able to remove this flag.

File: plugin/metrics/prometheus/options.go [hunk 2]
```diff
@@ -128,9 +125,6 @@ func (opt *Options) InitFromViper(v *viper.Viper) error {
 	cfg.TokenFilePath = v.GetString(cfg.namespace + suffixTokenFilePath)
 
 	cfg.SupportSpanmetricsConnector = v.GetBool(cfg.namespace + suffixSupportSpanmetricsConnector)
```
Comment (albertteoh) [inline plugin/metrics/prometheus/options.go:127]: I expected that we can remove the `SupportSpanmetricsConnector` config.

File: docker-compose/monitor/README.md [hunk 3]
```diff
@@ -321,3 +311,4 @@ $ curl http://localhost:16686/api/metrics/minstep | jq .
 [spanmetricsprocessor]: https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/processor/spanmetricsprocessor
 [spanmetricsconnector]: https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/connector/spanmetricsconnector
 [processor-to-connector]: https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/connector/spanmetricsconnector#span-to-metrics-processor-to-span-to-metrics-connector
```
Comment (albertteoh) [inline docker-compose/monitor/README.md:313]: We could probably remove this too and assume everyone's switched over to the connector.

File: docker-compose/monitor/README.md [hunk 4]
```diff
@@ -321,3 +311,4 @@ $ curl http://localhost:16686/api/metrics/minstep | jq .
 [spanmetricsprocessor]: https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/processor/spanmetricsprocessor
```
[the remark this answers is not in the document]
Comment (varshith257) [reply docker-compose/monitor/README.md:311]: okay

File: plugin/metrics/prometheus/options.go (hunk 1 above)
Comment (varshith257) [reply plugin/metrics/prometheus/options.go:95]: Will make changes

File: docker-compose/monitor/README.md (hunk 3 above)
Comment (varshith257) [reply docker-compose/monitor/README.md:313]: Can I also remove the migration guide too @albertteoh ?

**Comment by varshith257** [comment]
@albertteoh PTAL. I have also removed the migration guide. Would you also like me to update the configuration of the spanmetrics connector replacing the migration guide there?

File: docker-compose/monitor/README.md (hunk 3 above)
Comment (varshith257) [reply docker-compose/monitor/README.md]: Removed. Would you also like me to update the configuration of the spanmetrics connector replacing the migration guide there?

File: plugin/metrics/prometheus/options.go [hunk 5]
```diff
@@ -127,10 +118,6 @@ func (opt *Options) InitFromViper(v *viper.Viper) error {
 	cfg.ConnectTimeout = v.GetDuration(cfg.namespace + suffixConnectTimeout)
 	cfg.TokenFilePath = v.GetString(cfg.namespace + suffixTokenFilePath)
 
-	cfg.SupportSpanmetricsConnector = v.GetBool(cfg.namespace + suffixSupportSpanmetricsConnector)
```
Comment (albertteoh) [inline plugin/metrics/prometheus/options.go:130]: Thanks, I also meant that we can remove the actual config (and all its uses): https://github.com/jaegertracing/jaeger/blob/5742846bf691f0c0a88e9259720973270ac0c426/pkg/prometheus/config/config.go#L31

Comment (varshith257) [reply plugin/metrics/prometheus/options.go:130]: Done

File: plugin/metrics/prometheus/metricsstore/reader.go [hunk 6]
```diff
@@ -91,9 +91,6 @@ func NewMetricsReader(cfg config.Configuration, logger *zap.Logger, tracer trace
 	}
 
 	operationLabel := "operation"
-	if cfg.SupportSpanmetricsConnector {
-		operationLabel = "span_name"
```
Comment (albertteoh) [inline plugin/metrics/prometheus/metricsstore/reader.go:95]: The logic doesn't look right here.

File: plugin/metrics/prometheus/metricsstore/reader.go [hunk 7]
```diff
@@ -135,9 +132,6 @@ func (m MetricsReader) GetLatencies(ctx context.Context, requestParams *metricss
 
 func buildFullLatencyMetricName(cfg config.Configuration) string {
 	metricName := "latency"
```
Comment (albertteoh) [inline plugin/metrics/prometheus/metricsstore/reader.go:134]: This can be deleted.

File: plugin/metrics/prometheus/metricsstore/reader.go (hunk 6 above)
Comment (varshith257) [reply plugin/metrics/prometheus/metricsstore/reader.go:95]: What would be the appropriate option to be replaced there?

File: plugin/metrics/prometheus/metricsstore/reader.go (hunk 7 above)
Comment (varshith257) [reply plugin/metrics/prometheus/metricsstore/reader.go:134]: The latency histogram metric name was changed to duration as in OTEL migration guide. Can I replace with it?

File: plugin/metrics/prometheus/metricsstore/reader.go (hunk 6 above)
Comment (varshith257) [reply plugin/metrics/prometheus/metricsstore/reader.go:95]: The `operation` metric attribute was renamed to `span.name.`
Can be replaced with it?

Comment (albertteoh) [reply plugin/metrics/prometheus/metricsstore/reader.go:95]: Yup, I'd expect: `operationLabel := "span_name"`

File: plugin/metrics/prometheus/metricsstore/reader.go (hunk 7 above)
Comment (albertteoh) [reply plugin/metrics/prometheus/metricsstore/reader.go:134]: Yup, I expect: `metricName := "duration"`

Comment (varshith257) [reply plugin/metrics/prometheus/metricsstore/reader.go]: Done!

File: plugin/metrics/prometheus/metricsstore/reader.go (hunk 6 above)
Comment (varshith257) [reply plugin/metrics/prometheus/metricsstore/reader.go]: Done!

**Comment by albertteoh** [comment]
Looks like the tests need updating.

**Comment by varshith257** [comment]
Yea! Just fixed now

**Comment by varshith257** [comment]
@albertteoh The tests are failing as I update with the operation label as expected and converse it failing again with no need for an operation label :)

The current definitions of the functions `GetLatencies`, `GetCallRates,` and `GetErrorRates` in the reader.go, it seems these functions are designed to return `ErrDisabled` indicating that the metrics fetching functionality is disabled or not implemented

File: plugin/metrics/prometheus/metricsstore/reader_test.go [hunk 8]
```diff
@@ -217,17 +215,17 @@ func TestGetLatencies(t *testing.T) {
 			wantDescription: "0.95th quantile latency, grouped by service & operation",
 			wantLabels: map[string]string{
 				"service_name": "emailservice",
+				"operation":    "/OrderResult",
 			},
 			wantPromQlQuery: `histogram_quantile(0.95, sum(rate(span_metrics_duration_bucket{service_name =~ "emailservice", ` +
-				`span_kind =~ "SPAN_KIND_SERVER"}[10m])) by (service_name,span_name,le))`,
+				`span_kind =~ "SPAN_KIND_SERVER"}[10m])) by (service_name,operation,le))`,
```
Comment (albertteoh) [inline plugin/metrics/prometheus/metricsstore/reader_test.go:221]: We need to keep `span_name`. 

From: https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/connector/spanmetricsconnector/README.md#span-to-metrics-processor-to-span-to-metrics-connector

> The operation metric attribute was renamed to span.name.

I think we need to update this to `span_name` (and rename the file for consistency): https://github.com/jaegertracing/jaeger/blob/c0fb781662eacb4ec6f544e78d081528541c2473/plugin/metrics/prometheus/metricsstore/testdata/service_operation_datapoint_response.json#L8

So that we're mocking the metrics data response from the spanmetrics connector, which uses `span_name` instead of `operation`.

Comment (varshith257) [reply plugin/metrics/prometheus/metricsstore/reader_test.go:221]: Okay! will make changes

**Comment by varshith257** [comment]
@albertteoh It's 100% hit of test coverage now :)

**Comment by varshith257** [comment]
@yurishkuro PTAL and review the changes

File: plugin/metrics/prometheus/metricsstore/reader_test.go [hunk 9]
```diff
@@ -387,12 +380,11 @@ func TestGetErrorRates(t *testing.T) {
 			wantName:         "service_operation_error_rate",
 			wantDescription:  "error rate, computed as a fraction of errors/sec over calls/sec, grouped by service & operation",
 			wantLabels: map[string]string{
-				"operation":    "/OrderResult",
```
Comment (yurishkuro) [inline plugin/metrics/prometheus/metricsstore/reader_test.go:390]: why is this removed?

File: plugin/metrics/prometheus/metricsstore/reader_test.go [hunk 10]
```diff
@@ -286,11 +282,10 @@ func TestGetCallRates(t *testing.T) {
 			wantName:         "service_operation_call_rate",
 			wantDescription:  "calls/sec, grouped by service & operation",
 			wantLabels: map[string]string{
-				"operation":    "/OrderResult",
```
Comment (yurishkuro) [inline plugin/metrics/prometheus/metricsstore/reader_test.go:289]: why is this removed?

File: plugin/metrics/prometheus/metricsstore/reader_test.go (hunk 9 above)
Comment (albertteoh) [reply plugin/metrics/prometheus/metricsstore/reader_test.go:390]: This is a valid question @varshith257 . 

These two lines probably need updating: https://github.com/jaegertracing/jaeger/blob/22851437b082df712b2bc7e3da0e13ed222bf944/plugin/metrics/prometheus/metricsstore/reader_test.go#L903-L904

**Comment by varshith257** [comment]
@yurishkuro I have added back the var operation => span_name. It creates confusion about the already existing span_name variable there.  Isn't it `span.name`  instead span_name as mentioned in the migration guide 

`The operation metric attribute was renamed to span.name.
`

**Comment by albertteoh** [comment]
> Isn't it `span.name` instead span_name as mentioned in the migration guide

@varshith257  From: https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/pkg/translator/prometheus/README.md#labels

You'll see that one of the normalization steps performed to convert an OTLP attribute name to Prometheus is to: "Drop unsupported characters and replace with underscores (_)"

The example given there is to replace `host.name` -> `host_name`.

**Comment by varshith257** [comment]
The two lines of mentioned in  fn of mock Prometheus Server  have updated to span_name. After debugging the metricLabel in reader.go is still pointing to operation in tests though we already changed in reader go.

File: plugin/metrics/prometheus/metricsstore/reader_test.go [hunk 11]
```diff
@@ -311,13 +310,13 @@ func TestGetCallRates(t *testing.T) {
...
 			groupByOperation: true,
 			updateConfig: func(cfg config.Configuration) config.Configuration {
-				cfg.SupportSpanmetricsConnector = true
 				cfg.MetricNamespace = "span_metrics"
 				return cfg
 			},
 			wantName:        "service_operation_call_rate",
 			wantDescription: "calls/sec, grouped by service & operation",
 			wantLabels: map[string]string{
+				"span_name":    "/OrderResult",
```
Comment (yurishkuro) [inline plugin/metrics/prometheus/metricsstore/reader_test.go:319]: why are these added? It seems like it would change the semantics of the test.

cc @albertteoh

Comment (varshith257) [reply plugin/metrics/prometheus/metricsstore/reader_test.go:319]: In the original code, the test cases `Enable support for spanmetrics connector with a metric namespace` and `Enable support for spanmetrics connector with normalized metric name` did not include` span_name`(operation) in the `wantLabels `map. This test cases have the `groupByOperation` flag set to true, which means the Prometheus queries generated by the metrics system group by both service_name and span_name.

From the debug output, I confirmed that the metrics returned by these queries do indeed include the span_name label. This change ensures that the test cases accurately reflect the actual behavior when grouping. Including span_name in the wantLabels map ensures that the test cases validate the correct grouping and labeling of metrics. This change does not alter the semantics of the test but rather I believe it aligns the expected output with the current behavior.

If there is any scenario where grouping by operation is not intended, we can adjust the groupByOperation flag or the expected labels accordingly. However, for the current configurations, the inclusion of span_name is necessary and correct.

Comment (varshith257) [reply plugin/metrics/prometheus/metricsstore/reader_test.go:319]: Here's the  more debug output:

Test Name: group by service should be reflected in name/description and query group-by
Expected labels: map[service_name:emailservice]
Actual labels: [name:"service_name" value:"emailservice" ]
Remaining expected labels after matching: map[]
Test Name: group by service and operation should be reflected in name/description and query group-by
Expected labels: map[service_name:emailservice span_name:/OrderResult]
Actual labels: [name:"span_name" value:"/OrderResult"  name:"service_name" value:"emailservice" ]
Remaining expected labels after matching: map[]

Test Name: two services and span kinds result in regex 'or' symbol in query
Expected labels: map[service_name:emailservice]
Actual labels: [name:"service_name" value:"emailservice" ]
Remaining expected labels after matching: map[]

Test Name: enable support for spanmetrics connector with a namespace
Expected labels: map[service_name:emailservice span_name:/OrderResult]
Actual labels: [name:"span_name" value:"/OrderResult"  name:"service_name" value:"emailservice" ]
Remaining expected labels after matching: map[]

Test Name: enable support for spanmetrics connector with normalized metric name
Expected labels: map[service_name:emailservice span_name:/OrderResult]
Actual labels: [name:"span_name" value:"/OrderResult"  name:"service_name" value:"emailservice" ]
Remaining expected labels after matching: map[]

PASS
Process 23592 has exited with status 0

**Comment by varshith257** [comment]
@albertteoh PTAL

File: plugin/metrics/prometheus/metricsstore/reader_test.go (hunk 11 above)
Comment (albertteoh) [reply plugin/metrics/prometheus/metricsstore/reader_test.go:320]: Yeah, I think this is correct because the original mock spanmetrics connector JSON response (service_datapoint_response.json) didn't contain the `span_name`/`operation` label. 

Now we're using `service_span_name_datapoint_response.json` as the mock data which contains the `span_name` label.

**Review by albertteoh — APPROVED** [review]

**Comment by yurishkuro** [comment]
reverted changes to "operation" label #5673
